IPs/RSKIP144.md

```diff
@@ -9,48 +9,88 @@
+
+Values in `partitionsEnd` must be greater than 0 and in ascending order. An empty `partitionEnds` indicates that all transactions go in the sequential partition. The maximum number of parallel partitions that the miner can specify is equal to the minimum number of cores required to run the RSK node, this maximum is called  `maxTransactionExecutionThreads`.
+
+## New block validation consensus
+
+The new consensus mechanism allows to reject blocks that, when executed in parallel, can result in different outputs.
 
-When a miner executes transactions to create an (unsolved) block, the miner must execute the transactions serially and decide which partition the transaction belongs to. The block gas limit is replaced by a per-partition gas limit. We say that a transaction is *connected* to another transaction if:
+For simplicity, three types of connections between transactions are identified:
+- Both transactions are from the same sender account
+- Both transactions write the same storage key
+- One transaction reads a key that the other transaction writes
 
-1. Both transactions are from the same sender account
-2. Both transactions write the same storage key
-3. One transaction reads the same storage key that the other transaction writes
+Any pair of transactions with any of these characteristics are required to be in the same _parallel partition_ to avoid non-deterministic results, or one needs to be executed in the _sequential partition_
 
-A transaction is connected to a partition if it is connected to any transaction in that partition. The miner maintains a writeMap and a readMap that store which partition/s write or read each storage key. The miner then executes transactions from the pool and keeps track of which storage keys the transaction reads and writes. After executing the transaction, the miner compares the transaction read and written keys with the readMap and writeMap. The following scenarios are possible:
+This cases should not be considered invalid since they don't affect the output of the parallel execution:
+- Adding 0 balance
+- Writing a key and setting back the initial value in the same transaction
+- Creating a key and erasing it in the same transaction
+
+### Block validation algorithm
+
+A `readMap` and a `writeMap` is created per _parallel partition_. During execution of each partition,
+- when a storage key is read, it is marked in the `readMap`,
+- when a key is written, the key is marked in the `writeMap` map.
+
+When all parallel partitions have finished processing, the `readMap`s and `writeMap`s are scanned to find connections between transactions of different threads. This requires efficient maps that enable traversing the keys in ascending lexicographic order.
+- If a key belongs to a `writeMap` and a `readMap` of another _parallel partition_, then the block is considered invalid.
+- If a key belongs to a `writeMap` and a `writeMap` of another _parallel partition_, then the block is also considered invalid.
```

Comment:
This ended up being different in the implementation. We created two maps readKeys and writtenKeys that map key -> thread/s. 

When a thread writes a key it checks that no other thread has written the key, and that no other threads read the key.

Reply:
I will leave this one as it is until we review and merge the current implementation. In any case, this is a valid _suggested_ implementation.